src/lib/SaferNames/Inference.hs

```diff
@@ -1490,8 +1536,306 @@ freshType k = Var <$> freshInferenceName k
+  if hasInferenceVars bindings ty'
+    then
+      throw TypeErr "Can't synthesize a dictionary for a type with inference vars"
+    else do
+      let solutions = runSyntherM bindings $ buildBlock do
+                        ty'' <- injectM ty'
+                        synthDict ty''
+      let solutionsInstance = filter ((== UsedInstance) . snd) solutions
+      unless (length solutionsInstance <= 1) $
+        throw TypeErr $ "Multiple candidate class dictionaries for: " ++ pprint ty
+      case solutions of
+        [] -> throw TypeErr $ "Couldn't synthesize a class dictionary for: " ++ pprint ty
+        (d, _):_ -> injectM d
+
+-- TODO: we'd rather have something like this:
+--   data Givens n = Givens (M.Set (Type n) (Given n))
+-- but we need an Ord or Hashable instance on types
+data Givens n = Givens [Type n] [Given n]
+
+type Given = Block
+
+data DerivationKind = OnlyGivens | UsedInstance deriving Eq
+
+instance Semigroup DerivationKind where
+  OnlyGivens <> OnlyGivens = OnlyGivens
+  _          <> _          = UsedInstance
+instance Monoid DerivationKind where
+  mempty = OnlyGivens
+
+class (Alternative1 m, Searcher1 m, Builder m)
+      => Synther m where
+  getGivens :: m n (Givens n)
+  withGivens :: Givens n -> m n a -> m n a
+  declareUsedInstance :: m n ()
+  -- TODO: just expose `unsafeGetBindings` etc to avoid this silly rank-2 dance
+  liftSolverSynth
+    :: (SubstE AtomSubstVal e', HoistableE e', InjectableE e, InjectableE e')
+    => e n
+    -> (forall l. e l -> SolverM l (e' l))
+    -> m n (e' n)
+
+newtype SyntherM (n::S) (a:: *) = SyntherM
+  { runSyntherM' :: OutReaderT Givens (BuilderT (WriterT DerivationKind [])) n a }
+  deriving ( Functor, Applicative, Monad, BindingsReader
+           , Scopable, ScopeReader, MonadFail
+           , Alternative, Searcher, OutReader Givens)
+
+instance Synther SyntherM where
+  getGivens = askOutReader
+
+  declareUsedInstance = SyntherM $ tell UsedInstance
+
+  withGivens givens cont = do
+    localOutReader givens cont
+
+  liftSolverSynth e cont = do
+    WithBindings bindings e' <- addBindings e
+    let result = runSolverM bindings do
+                   e'' <- injectM e'
+                   cont e''
+    case result of
+      Success result' -> injectM result'
+      Failure _ -> empty
+
+instance Builder SyntherM where
+  emitDecl hint ann expr = SyntherM $ emitDecl hint ann expr
+  buildScopedGeneral ab f = SyntherM $
+    buildScopedGeneral ab \e ->
+      runSyntherM' $ f e
+
+runSyntherM :: Distinct n
+            => Bindings n
+            -> (forall l. (Distinct l, Ext n l) => SyntherM l (e l))
+            -> [(e n, DerivationKind)]
+runSyntherM bindings cont = do
+  runWriterT $ runBuilderT bindings do
+    initGivens <- givensFromBindings
+    runOutReaderT initGivens $ runSyntherM' cont
+
+givensFromBindings :: BindingsReader m => m n (Givens n)
+givensFromBindings = do
+  WithBindings bindings UnitE <- addBindings UnitE
```

Comment:
what does this do?

Reply:
Just gets the givens indicated by the static env (what we currently call `Bindings`, though it's now a misnomer). The inference monad tracks its own givens (since higher-order constraints introduce new givens) so we only use the static env for initialization.